R/cell_suppression.R

```diff
@@ -252,103 +255,114 @@ suppress_secondary <- function(
+          # Successful solution
+          alpha <- attacker.min$solution[seq_len(n)]
+          beta <- attacker.min$solution[n + seq_len(n)]
+
+          constraints(master_lp) <- c(
+            constraints(master_lp),
+            # ref, eqn (26)
+            L_constraint(
+              pmin(alpha * UB + beta * LB, LPL[attack.ik]), # Sec 4.2.1
+              dir = '>=',
+              rhs = LPL[attack.ik]
+            )
           )
-        )
-
-        attacker_success <- TRUE
-      }
-
-      if (SPL[attack.ik] > 0L &&
-          identical(attacker.max$status$code, 0L) &&
-          identical(attacker.min$status$code, 0L) &&
-          # Note negation in eqn (23)
-          attacker.max$objval + attacker.min$objval < SPL[attack.ik]) {
-
-        alpha <- attacker.min$solution[seq_len(n)] +
-          attacker.max$solution[seq_len(n)]
-
-        beta <- attacker.min$solution[n + seq_len(n)] +
-          attacker.max$solution[n + seq_len(n)]
-
-        constraints(master_lp) <- c(
-          constraints(master_lp),
-          # ref, eqn (27)
-          L_constraint(
-            pmin(alpha * UB + beta * LB, SPL[attack.ik]), # Sec 4.2.1
-            dir = '>=',
-            rhs = SPL[attack.ik]
+
+          attacker_success <- TRUE
+        }
+
+        # Evaluate SPL subproblem
+        if (SPL[attack.ik] > 0L &&
+            identical(attacker.max$status$code, 0L) &&
+            identical(attacker.min$status$code, 0L) &&
+            # Note negation in eqn (23)
+            attacker.max$objval + attacker.min$objval < SPL[attack.ik]) {
+
+          alpha <- attacker.min$solution[seq_len(n)] +
+            attacker.max$solution[seq_len(n)]
+
+          beta <- attacker.min$solution[n + seq_len(n)] +
+            attacker.max$solution[n + seq_len(n)]
+
+          constraints(master_lp) <- c(
+            constraints(master_lp),
+            # ref, eqn (27)
+            L_constraint(
+              pmin(alpha * UB + beta * LB, SPL[attack.ik]), # Sec 4.2.1
+              dir = '>=',
+              rhs = SPL[attack.ik]
+            )
           )
-        )
 
-        attacker_success <- TRUE
+          attacker_success <- TRUE
+
+        }
 
-      }
+    }
 
-      # If there is no successful avenue of attack, move to the next suppressed cell
-      if (!attacker_success) break
+    # If there is no successful avenue of attack, calculation complete
+    if (!attacker_success) break
 
-      master_soln <- ROI_solve(master_lp, ...)
+    # Exit with error if iteration limit exceeded
+    if (identical(i, max_iter))
+      stop("Maximum attacker iterations reached. Consider increasing the max_iter parameter")
 
-      candidate_suppression <- as.logical(master_soln$solution)
-    }
+    # Re-solve master LP with new constraints to feed next cycle
+    i <- i + 1L
+    candidate_suppression <- as.logical(ROI_solve(master_lp, ...)$solution)
+    candidate_suppression <- ROI_solve(master_lp, ...)$solution >= 0.5
```

Comment:
We should probably save the master LP solution as an object and check the status code before running with the new candidate suppression.

(Also this looks like it might be a duplicate line.)

Reply:
🤦‍♂️ You are right, that is leftover from doing testing. Have deleted the `as.logical()` line. Good idea on checking status code before running new `candidate_suppression`. I have added an error that passes the message from the solver back to the user if status code != 0 as without running a new `candidate_suppression`, it will otherwise loop back and continue tripping the same subproblems until it hits `max_iter`. Would appreciate your thoughts on any situation where a non-0 status code would still be appropriate to continue solving